### ISSUE TITLE
Add HTTP headers

### ISSUE THREAD

**Issue by KINGSABRI** [sdaubert/packetgen#59]

**Comment by sdaubert** [comment]
PacketGen targets at parsing and generating binary protocols. And there is already plenty of ruby  HTML parser which do the job, and better than PacketGen will.

So, this will not be done.

**Comment by picatz** [comment]
@sdaubert Forgive me if I don't know what I'm talking about: but, I think targeting something like **[HTTP 1.1](https://developer.mozilla.org/en-US/docs/Web/HTTP/Overview)** would be an awesome feature for packetgen! 

I could even try to implement it myself. Not terribly sure how successful I'd be / don't know everything about it. But I'd love this feature which some other parsers like [nDPI](https://github.com/ntop/nDPI/blob/8bcafff56991714bad24cb4d788e26888c25df70/src/lib/protocols/http.c) and [Wireshark](https://wiki.wireshark.org/Hyper_Text_Transfer_Protocol) have implemented to some extent.

#### Simple HTTP 1.1 Parser [Example](https://stackoverflow.com/questions/13289268/udp-http-header-parsing-in-ruby)
```ruby
# very rough test, maybe won't get you an HTTP packet that parseable if it's HTTPs, right?
http_packet = PacketGen.capture(iface: 'eth0', filter: 'tcp and dst port 80', max: 1)

# http header information to_hash
http_header = http_packet.body.strip.split($/).drop(1).inject({}) do |h, l|
  k, v = l.split(": ", 2)
  h[k.downcase.gsub("-", "_").to_sym] = (v.strip unless v.nil? || v.empty?)
  h
end

http_header[:host]
# =>"www.tcpdump.org"

http_header[:user_agent]
# => "I like waffles, and ruby!."
```

_I understand if you still don't want to do this._ 😅  But I figure I'd at least say I'd like this feature. Don't think we need to deal with parsing HTML or JSON or XML for people since that's usually what the http headers are helping transport. But, it'd be nice to easily identify http packets / have shortcut methods to its headers.

I appreciate all your efforts on the protocols you've implemented thus far! 👍

**Comment by sdaubert** [comment]
@picatz If you want to do it, don't hesitate. I rejected this proposal because I don't want to do it, but if someone else wants, that's not a problem.

But, if you do it, don't forget to also make generator part.

And, please, don't forget specs.

What do you think about that?

**Comment by picatz** [comment]
@sdaubert This could be a good compromise. 👍 

Truthfully, I am not fully confident in how to implement this yet. 

I will probably have some more time to work on this tonight / tomorrow. I will have to go over your previous commits of adding other protocols to see how it works. And [then go back over the wiki](https://github.com/sdaubert/packetgen/wiki/Create-Custom-Protocol) to understand adding a custom protocol. It will be good experience to dive into it more.

Hopefully I ( or @KINGSABRI 😉 ) can figure this out, it'd be really cool!

**Comment by sdaubert** [comment]
@picatz the wiki won't help, as it explains how to implement a binary protocol. I thin you will have to start from a blank sheet.

Nevertheless, you have to implement a minimal set of methods to be compatible with `Packet#parse`:
* `packet` attribute,
* `.protocol_name` and `#protocol_name`,
* `#method_name`,
* `#parse?`,
* `#packet=`, only if you need `#added_to_packet`,
* `#read`, which really decodes header.

You may have a look at [Header::ASN1Base](http://www.rubydoc.info/gems/packetgen/PacketGen/Header/ASN1Base).

You also have to implement `#to_s` to encode header.

I think your big job will be to write `#read` and `#to_s`, others methods should be copied from `Header::Base` or `Header::ASN1Base`.

**Comment by picatz** [comment]
@sdaubert I am working on the HTTP 1.1 implementation today. 

Right now I've got a **_really simple_** implementation going like so:

```ruby
module PacketGen
  module Header

    # An HTTP/1.1 packet.
    # @author Kent 'picat' Gruber
    class HTTP < Base
      define_field :all, Types::String
    
      def read(str)
        binding.pry
        self[:all] = str 
        self
      end

    end

    self.add_class HTTP 
    TCP.bind_header HTTP, dport: 80, sport: 80
  end
end
```

I understand this **isn't done by any means**. But, I'm working on it 👍 

## Binding the HTTP Header
I think I need to figure out how to bind to a TCP header correctly. 

Because right now I'll end up picking up non-HTTP (TCP) requests leading up to the actual HTTP request or something like that. I think I need a ProcBinding to sort this out? 

I dunno, but I'll figure it out 😹

**Comment by picatz** [comment]
Playing around with it some more -- hopefully all today. But I wanted to give an update just in case @KINGSABRI also wanted to work on this, so I wanted to share:

### HTTP REQUEST 
This seems to be a decent-ish, simple [HTTP Request](https://www.w3.org/Protocols/rfc2616/rfc2616-sec5.html) TCP header binding using a Proc.
```ruby
TCP.bind_header HTTP, body: ->(b) { /^(CONNECT|DELETE|GET|HEAD|OPTIONS|PATCH|POST|PUT)/ =~ b }
```

**Comment by picatz** [comment]
And, _for just a bit of clarification_, I do plan on breaking down the actual HTTP request/response headers. It would probably be to end up breaking them up into their respective types instead of just one `http` header type?

I'm just using the `self[:all] = str` for the `#read` method to act as a simple catch-all because it's easier to do that for now.

I'm also not terribly sure on what to name the header. Maybe just call it `HTTP` and then, if http/2 was needed/added later, it could be `HTTP2`? 🤔 

( _Sorry if these thoughts are a little scattered, since I'm still forming them myself._ )

**Comment by sdaubert** [comment]
```ruby
TCP.bind_header HTTP, dport: 80, sport: 80
```
That should work, i.e. it should parse a HTTP header when TCP dport or sport is set to 80.

> I do plan on breaking down the actual HTTP request/response headers.

You may create a base HTTP class which demultiplex on parsing to specialized classes (by example HTTP::Request and HTTP::Response). Have a look at EAP or Dot11 for examples.

**Comment by picatz** [comment]
@sdaubert Maybe this would be better for the request header type:

```ruby
# when dport is port 80 and body is likely parseable
TCP.bind_header HTTP, , dport: 80, body: ->(b) { /^(CONNECT|DELETE|GET|HEAD|OPTIONS|PATCH|POST|PUT)/ =~ b }
```

I don't want to catch all of the stuff that hits port 80 since some of it isn't sending any HTTP data to parse, just doing your typical TCP stuff to connect from what I can tell.

## HTTP Demultiplex 
 I'll try to follow what you did with `EAP` or `Dot11`, like you said. 👍

**Comment by picatz** [comment]
@sdaubert, @KINGSABRI I think you're gunna like this, even if I'm not done:

```ruby
---- PacketGen::Header::HTTP::Request --------------------------------
          String       method: "GET"
          String         path: "/manpages/pcap_get_selectable_fd.3pcap.txt"
          String      version: "HTTP/1.1"
      HTTPHeaders      headers: {:host=>"www.tcpdump.org", :user_agent=>"android-async-http/1.4.4 (http://loopj.com/android-async-http)", :accept=>"*/*"}
```

I like the idea of presenting the headers as a hash. It feels the most natural thing to do in pry.

Also...

```ruby
packet.http_request.to_s
# => "GET /manpages/pcap_get_selectable_fd.3pcap.txtHTTP/1.1\r\nHost: www.tcpdump.org\r\nUser-Agent: android-async-http/1.4.4 (http://loopj.com/android-async-http)\r\nAccept: */*\r\n\r\n"
```

So, I'm thinking I'm going to be able to do this! 🤘

**Comment by picatz** [comment]
#### Unnecessary Update
* Fight for 20 minutes with my code. 😟 
* Realize my `bpf` filter is set to just look for outgoing requests. 📤 
* Realize I should have a snack. 🇭🇺 
* Haz snacked. 🍬 
* Ready to get the `HTTP::Response` class knocked out. 😼 
* Should actually `commit` something to my fork. 🤔 
* Yolo 😹

**Comment by picatz** [comment]
@sdaubert How do we ( or, **_can_** we ) handle streaming responses in PacketGen?

I can do something like just match for the first portion of a streaming HTTP response from a server. But, if I want to identify each following packet in the stream, is there an easy way to do this?

**Comment by picatz** [comment]
I guess don't really mind....

 But, it _won't_ serialize **everything** in a streaming HTTP response. I guess it would require PacketGen to be something like a _stateful_ packet parser and be able to follow the different streams that can happen when parsing -- which could be out of scope / or be a separate project maybe?

## HTTP Response Example
```ruby
---- PacketGen::Header::HTTP::Response -------------------------------
          String      version: "HTTP/1.1"
          String  status_code: "200"
          String  status_mesg: "OK"
      HTTPHeaders      headers: {:date=>"Sun, 12 Nov 2017 23:14:02 GMT", :server=>"Apache/2.4.10 (Debian)", :last_modified=>"Sun, 22 Jan 2017 06:52:58 GMT", :etag=>"\"d17-546a9554bc680\"", :accept_ranges=>"bytes", :content_length=>"3351", :content_type=>"text/plain"}
          String         data: "PCAP_GET_SELECTABLE_FD(3PCAP)                    PCAP_GET_SELECTABLE_FD(3PCAP)\n\nNAME\n       pcap_get_selectable_fd  - get a file descriptor on which a select() can\n       be done for a live capture\n\nSYNOPSIS\n       #include <pcap/pcap.h>\n\n       int pcap_get_selectable_fd(pcap_t *p);\n\nDESCRIPTION\n       pcap_get_selectable_fd() returns, on UNIX, a file descriptor number for\n       a file descriptor on which one can do a select(), poll(), or other such\n       call to wait for it to be possible to read packets without blocking, if\n       such  a  descriptor  exists, or -1, if no such descriptor exists.  Some\n       network devices opened with pcap_create() and pcap_activate(), or  with\n       pcap_open_live(), do not support select() or poll() (for example, regu-\n       lar network devices on FreeBSD 4.3 and 4.4, and Endace DAG devices), so\n       -1 is returned for those devices.\n\n       Note  that  a  descriptor  on which a read can be done without blocking\n       may, on some platforms, not have any packets to read if the packet buf-\n       fer  timeout  has  expired.  A call to pcap_dispatch() will return 0 in\n       this case, but will not block.\n\n       Note that in:\n\n              FreeBSD prio"
```

**Comment by picatz** [comment]
Yooooo I think I did itttttt.

## HTTP Request Example
```ruby
---- PacketGen::Header::HTTP::Request --------------------------------
          String       method: "GET"
          String         path: "/manpages/pcap_get_selectable_fd.3pcap.txt"
          String      version: "HTTP/1.1"
      HTTPHeaders      headers: {:host=>"www.tcpdump.org", :user_agent=>"android-async-http/1.4.4 (http://loopj.com/android-async-http)", :accept=>"*/*"}
```

## HTTP Response Example
```ruby
---- PacketGen::Header::HTTP::Response -------------------------------
          String      version: "HTTP/1.1"
          String  status_code: "200"
          String  status_mesg: "OK"
      HTTPHeaders      headers: {:date=>"Mon, 13 Nov 2017 03:03:41 GMT", :server=>"Apache/2.4.10 (Debian)", :last_modified=>"Sun, 22 Jan 2017 06:52:58 GMT", :etag=>"\"d17-546a9554bc680\"", :accept_ranges=>"bytes", :content_length=>"3351", :content_type=>"text/plain"}
---- Body ------------------------------------------------------------
 00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15
----------------------------------------------------------------------
 50 43 41 50 5f 47 45 54 5f 53 45 4c 45 43 54 41  PCAP_GET_SELECTA
 42 4c 45 5f 46 44 28 33 50 43 41 50 29 20 20 20  BLE_FD(3PCAP)   
 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20                  
 20 50 43 41 50 5f 47 45 54 5f 53 45 4c 45 43 54   PCAP_GET_SELECT
 41 42 4c 45 5f 46 44 28 33 50 43 41 50 29 0a 0a  ABLE_FD(3PCAP)..
....
```

**Comment by picatz** [comment]
Working on the PR now to hopefully ensure it merges without too many problems.

```ruby
# capture some http packets
http_packets = PacketGen.capture(iface: 'eth0', filter: 'tcp and ip host 192.139.46.66 and port 80', max: 10)
```
For a simple test:
```shell
curl http://www.tcpdump.org/manpages/pcap_get_selectable_fd.3pcap.txt
```

Get yo'self a `HTTP::Request` header.
```ruby
packet = http_packets.find { |p| p.http_request }
# check data
packet.http_request.version
# => "HTTP/1.1"
packet.http_request.method
# => "GET"
packet.http_request.path
# => "/manpages/pcap_get_selectable_fd.3pcap.txt"
packet.http_request.headers
# => {:host=>"www.tcpdump.org", :user_agent=>"android-async-http/1.4.4 (http://loopj.com/android-async-http)", :accept=>"*/*"}
# edit the data?
packet.http_request.method = "PUT"
# http headers as hash
packet.http_request.headers[:host] = "www.google.com"
``` 

Get yo'self a `HTTP::Response` header.
```ruby
packet = http_packets.find { |p| p.http_response }
# check data
packet.http_response.version
# => "HTTP/1.1"
packet.http_response.status_code
# => "200"
packet.http_response.status_code
# => "OK"
packet.http_request.headers
# => {:date=>"Mon, 13 Nov 2017 03:03:41 GMT", :server=>"Apache/2.4.10 (Debian)", :last_modified=>"Sun, 22 Jan 2017 06:52:58 GMT", :etag=>"\"d17-546a9554bc680\"", :accept_ranges=>"bytes", :content_length=>"3351", :content_type=>"text/plain"}
```

**Comment by picatz** [comment]
Will go about fixing the problems @sdaubert pointed out in my PR https://github.com/sdaubert/packetgen/pull/64 as soon as I can. Hopefully by this weekend. 👍

**Comment by picatz** [comment]
Working on the problems today. Hopefully I can get them all sorted out. 👀

**Comment by picatz** [comment]
To keep this section updated as well, for whoever may be interested. Closed https://github.com/sdaubert/packetgen/pull/64 and then opened https://github.com/sdaubert/packetgen/pull/65 🙈

**Comment by sdaubert** [comment]
Fixed by #65
